### PR TITLE
Query Selenium Server to retrieve actual running port.

### LINKED ISSUE #1299: Incorrect handling of "port 0" in Selenium Grid setup
###### Background:
* Under Linux and Windows, bind()-ing to network port 0 assigns the application a free port from the dynamic port range.
* I opened a similar ticket (#1292) but it was rejected because I only referred to deprecated RC tests, but the described behaviour also affects WebDriver tests.

###### Test setup:
* Selenium server standalone version 2.48.2
* CentOS 7
* Java 1.7.0_91

###### Issue description:
* In a Selenium Grid setup, nodes launched using option `-port 0` will correctly listen to a random free port assigned by the OS but will wrongly register on the hub using **port 0** instead of the effective port they are listening on.

###### Example:
While having a hub listening on port 4444, launch a node with a `nodeconfig.json` similar to:
```
{
  "capabilities":
      [
        {
          "browserName": "firefox" ,
          "maxInstances": 2,
          "seleniumProtocol": "WebDriver"
        }
      ],
  "configuration":
  {
    "proxy": "org.openqa.grid.selenium.proxy.DefaultRemoteProxy",
    "maxSession": 2,
    "port": 0,
    "host": 127.0.0.1,
    "register": true,
    "registerCycle": 5000,
    "hubPort": 4444,
    "hubHost": 127.0.0.1
  }
}
```

`lsof` will show that the node process listens on a random port offered by the OS, but the Grid console will show the node as being registered with **port 0** instead of the effective listening port:
```
nodeConfig:nodeconfig3.json
port:0
hubConfig:/opt/selenium/hubconfig.json
servlets:[]
host:127.0.0.1
cleanUpCycle:5000
browserTimeout:0
hubHost:127.0.0.1
registerCycle:5000
capabilityMatcher:org.openqa.grid.internal.utils.DefaultCapabilityMatcher
newSessionWaitTimeout:-1
url:http://127.0.0.1:0
remoteHost:http://127.0.0.1:0
prioritizer:null
register:true
throwOnCapabilityNotPresent:true
nodePolling:5000
proxy:org.openqa.grid.selenium.proxy.DefaultRemoteProxy
maxSession:2
role:node
jettyMaxThreads:-1
hubPort:4444
timeout:300000 
```
* No submitted Webdriver, RC, etc. tests will be able to run on this wrongly registered node.
 
Also this incorrect behaviour seems to affect all the possible use cases of Selenium (see also the closed ticket #1292 )

Kind regards,

Volker

### PR DESCRIPTION
**Pull request by krmahadevan**
- [ X] By placing an `X` in the preceding checkbox, I verify that I have signed the [Contributor License Agreement](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md#step-6-sign-the-cla)

Fixes #1299

Altered the registration logic as below :
If port given was 0, then query server to find
running port and update registration request with 
correct values for remoteHost.

### CODE REVIEW

**Comment by krmahadevan** [comment]
@lukeis 
If time permits, was wondering if you could please help me with this PR

File: java/server/src/org/openqa/grid/internal/utils/SelfRegisteringRemote.java [hunk 1]
```diff
@@ -216,6 +216,15 @@ private void registerToHub(boolean checkPresenceFirst) {
 
         BasicHttpEntityEnclosingRequest r =
             new BasicHttpEntityEnclosingRequest("POST", registration.toExternalForm());
+        int port = Integer.parseInt(nodeConfig.getConfigAsString(RegistrationRequest.PORT));
+        if (port == 0) {
+          port = server.getRealPort();
+          Map<String, Object> config = nodeConfig.getConfiguration();
+          config.put(RegistrationRequest.PORT,server.getRealPort());
+          URL url = new URL(nodeConfig.getConfigAsString(RegistrationRequest.REMOTE_HOST));
+          String newUrl = "http://" + url.getHost() + ":" + port;
```
Comment (lukeis) [inline java/server/src/org/openqa/grid/internal/utils/SelfRegisteringRemote.java]: don't assume "http"  use url.getProtocol()

**Comment by lukeis** [comment]
in general looks good... but would really be great to have a test checked in for this

**Comment by krmahadevan** [comment]
@lukeis 
Thank you so much for helping with the review. I have fixed your review comments and also added unit tests as you suggested.

**Comment by krmahadevan** [comment]
@lukeis 
I will fix the compilation failure and update this PR [ still haven't got a hang of the selenium build system.. so please bear with me ]

**Comment by krmahadevan** [comment]
@lukeis 
I have fixed the problem. Can you please see if all is well with this PR and if it can be accepted ?

**Comment by krmahadevan** [comment]
@lukeis 
Looks like you have been busy with Grid3 :)
Was wondering if you had some time for this PR...

**Comment by lukeis** [comment]
I'll try to get it merged soon :)

**Comment by lukeis** [comment]
merged as 31a991475538f496cd65d68c379cf9a3be3aa03d

**Pull request closed without merging**